cantabular/table.go

```diff
@@ -96,6 +129,60 @@ func GraphQLJSONToCSV(ctx context.Context, r io.Reader, w io.Writer) (int32, err
 	return rowCount, nil
 }
 
+// GraphQLJSONToJson converts a JSON response in r to a different schema on w, returning the response
+// if an error happens, the process is aborted and the error is returned.
+func GraphQLJSONToJson(ctx context.Context, r io.Reader, w io.Writer) (GetObservationsResponse, error) {
+	dec := jsonstream.New(r)
+
+	// errData represents a possible error that may be returned by 'decodeDataFields',
+	// as long as it is reported in 'errors'
+	var errData error
+
+	var getObservationsResponse GetObservationsResponse
+
+	// find starting '{'
+	if isStartObj, err := dec.StartObjectComposite(); err != nil {
+		return GetObservationsResponse{}, fmt.Errorf("error decoding start of json object: %w", err)
+	} else if !isStartObj {
+		return GetObservationsResponse{}, errors.New("no json object found in response")
+	}
+
+	// decode 'data' and 'error' fields
+	for dec.More() {
+		field, err := dec.DecodeName()
+		if err != nil {
+			return GetObservationsResponse{}, fmt.Errorf("error decoding field: %w", err)
+		}
+		switch field {
+		case "data":
+			if getObservationsResponse, err = decodeDataFieldsJson(ctx, dec, w); err != nil {
+				// null values for 'dataset' or 'table' are ok as long as the error is reported under the 'errors' field
```

Comment:
I'm not entirely clear on how this functionality is working - It seems to be that if it returns an error here, the loop stops being executed and then we go onto the `EndComposite` and checking `errData` !=nil where it will error. Maybe I'm missing something? I can see it's been implemented in this way in other places in this file.

Reply:
This is taken from code that SCC provided for us a while ago so my very basic understanding of it is that if the fields dataset or table are returned from the Cantabular graphQL query then there should be something in the errors field instead - an example response might look something like this in this case:

{
  "data": {
    "dataset": {
      "table": null
    }
  },
  "errors": [
    {
      "message": "400 Bad Request: variable at position 2 does not exist",
      "locations": [
        {
          "line": 2,
          "column": 23
        }
      ],
      "path": [
        "dataset",
        "table"
      ]
    }
  ]
}